### PR TITLE
feat: support for (mutual) recursion in malli.json-schema

### LINKED ISSUE #464: Missing "Defs" in swagger (malli.json-schema) with recursive schemes
Hi! I try to use malli as coercion&schemas with swagger ui. If i try serialise recursive schemas with registries, i get some "Could not resolve reference" of keys of recursive values. it is according to plan and swagger bad, or its just undone feature?) looks like there is some missing defs in swagger.json
image

I use swagger-ui bundled with

`[metosin/reitit-swagger-ui "0.5.13"]
[metosin/ring-swagger-ui "3.36.0"]`

guess it 3.XX.

I'v got example of that malli:

```
[:schema
 {:registry {:namespace1/logic-pred [:or
                                        [:= #:json-schema{:val "and"} "and"]
                                        [:= #:json-schema{:val "or"} "or"]],
             :namespace1/not-pred [:= #:json-schema{:val "not"} "not"],
             :namespace1/fields-or-expression [:or
                                                [:ref "Expression"]
                                                [:or
                                                 [:tuple
                                                  [:= #:json-schema{:val "="} "="]
                                                  [:= #:json-schema{:example "phone"} "phone"]
                                                  string?]
                                                 [:tuple
                                                  [:= #:json-schema{:val "contains"} "contains"]
                                                  [:= #:json-schema{:example "phone"} "phone"]
                                                  string?]
                                                 [:tuple
                                                  [:= #:json-schema{:val "startswith"} "startswith"]
                                                  [:= #:json-schema{:example "phone"} "phone"]
                                                  string?]
                                                 [:tuple
                                                  [:= #:json-schema{:val "like"} "like"]
                                                  [:= #:json-schema{:example "phone"} "phone"]
                                                  string?]]
                                                
                                                ;; ... (more :or's)
                                                
                                                                  ],
             "Expression" [:cat
                           :namespace1/logic-pred
                           [:+
                            [:or
                             [:ref :namespace1/fields-or-expression]
                             [:tuple
                              :namespace1/not-pred
                              [:ref :namespace1/fields-or-expression]]]]]}}
 [:map
  [:where {:optional true, :json-schema/val "[and [= a 2]]"} "Expression"]
  [:limit [int? {:default 100}]]
  [:offset [int? {:default 0}]]]]
```

If i malli.json-schema/transform  this (it is used in swagger gen, right?) i get something like:

```
{:type "object",
 :properties {:where {:$ref "#/definitions/Expression", :val "[and [= a 2]]"},
              :limit {:default 100, :type "integer"},
              :offset {:default 0, :type "integer"}},
 :required [:limit :offset],
 :definitions {:namespace1/logic-pred {:anyOf [{:const "and", :val "and"} {:const "or", :val "or"}]},
               :namespace1/not-pred {:const "not", :val "not"},
               "Expression" {}}}
```

and it has no valuable
`#/definitions/Expression`
as you can see

### LINKED ISSUE #884: mu/closed-schema (and other m/schema-walker uses) produce extra wrappers for reference schemas
The result of `mu/closed-schema` needs one more `m/deref` to get to the concrete schema. The same happens with a trivial `m/schema-walker` use.

Demonstration:
```clj
user=> (require '[malli.core :as m])
nil
user=> (require '[malli.util :as mu])
nil
user=> (def s1 (m/schema [:schema {:registry {"Foo" :int}} "Foo"]))
#'user/s1
user=> (def s2 (mu/closed-schema s1))
#'user/s2
user=> (m/form s1)
[:schema {:registry {"Foo" :int}} "Foo"]
user=> (m/form s2)
[:schema {:registry {"Foo" :int}} "Foo"]
user=> (-> s1 m/deref m/deref)
:int
user=> (-> s2 m/deref m/deref)
"Foo"
user=> (-> s2 m/deref m/deref m/deref)     ;; one more deref needed!
:int
user=> (defn noop-walk [schema] (m/walk schema (m/schema-walker identity)))
#'user/noop-walk
user=> (def s3 (noop-walk s1))
#'user/s3
user=> (-> s3 m/deref m/deref)
"Foo"
user=> (-> s3 m/deref m/deref m/deref)     ;; one more deref needed!
:int
```

The root cause is the `-walk` implementation for `-schema-schema`:

```clj
            (-walk [this walker path options]
              (when (-accept walker this path options)
                (if (or (not id) ((-boolean-fn (::walk-schema-refs options false)) id))
                  (-outer walker this path (-inner-indexed walker path children options) options)
                  (-outer walker this path [id] options))))
```

This combined with `m/schema-walker` causes `(m/-set-children schema [id])`, which overrides the child from `:int` to `"Foo"`. Luckily this doesn't cause a loop in later schema walking, just the need for one more deref.

### PR DESCRIPTION
**Pull request by opqdonut**
Mutual recursion support. Fixes #464 #868

A schema walker bug was making life difficult so this includes a fix for it. Fixes #884 . Removed `mu/from-map-syntax` and `mu/to-map-syntax` to make the fix easier.

### CODE REVIEW

File: src/malli/util.cljc [hunk 1]
```diff
@@ -120,7 +120,9 @@
 
   See [[open-schema]]"
   ([?schema]
-   (closed-schema ?schema nil))
+   ;; We default to ::m/walk-schema-refs true because that avoids
+   ;; breaking the children of pointer schemas. See closed-pointer-schema-regression-test.
+   (closed-schema ?schema {::m/walk-schema-refs true}))
```
Comment (opqdonut) [inline src/malli/util.cljc:125]: I don't really like this fix, but it makes everything work out nicely.

**Comment by opqdonut** [comment]
This fix has some weird consequences:

```clj
user> (def closed (mu/closed-schema [:schema {:registry {"Foo" [:map [:a :int]]}} "Foo"]))
#'user/closed
user> (m/form closed)
[:schema {:registry {"Foo" [:map [:a :int]]}} "Foo"]
user> (-> closed m/deref m/deref m/form)
[:map {:closed true} [:a :int]]
```

So the instance of `"Foo"` used at the top level is mutated (and now differs from the one in the registry), but it doesn't show up in the `m/form` because it's hidden by the `id` of the `:schema`.

**Comment by opqdonut** [comment]
I think I'll have to rethink the closed-schema fix, perhaps on the level of schema-walker or set-children.

**Comment by opqdonut** [comment]
I managed to implement a workaround for malli.json-schema. I opened #884 for the mu/closed-schema problem. This PR is mergeable even if we don't solve #884.

**Comment by opqdonut** [comment]
I think I have an actual solution to the walker problem now. I'll test it out & then rebase this PR to get rid of all the intermediate steps.

**Comment by opqdonut** [comment]
just added a test on the schema-walker level

File: src/malli/core.cljc [hunk 2]
```diff
@@ -2360,8 +2360,15 @@
 ;;
 
 (defn schema-walker [f]
-  (fn [schema _ children _]
-    (f (-set-children schema children))))
+  (fn [schema _ children options]
+    ;; Workaround for #884. The -walk method of -schema-schema passes
+    ;; in children=[id], which would cause us to (-set-children s [id]),
+    ;; overwriting the actual child.
+    (f (if (and (#{:schema ::schema} (-> schema -parent -type))
```
[the remark this answers is not in the document]
Comment (opqdonut) [reply src/malli/core.cljc:2367]: Discussed this and created #893 .

**Comment by opqdonut** [comment]
I tried removing the `[id]` branch from `-schema-schema`. That caused a number of tests to fail, so we are relying on the `[id]` case. Those could probably be reworked, but we might break outsider users as well. However, all my new tests passed, so this is a viable fix.

If you prefer, I can go forward with this slightly-nicer but slightly-more-breaking fix.

```diff
diff --git a/src/malli/core.cljc b/src/malli/core.cljc
index aa21916..6d53e20 100644
--- a/src/malli/core.cljc
+++ b/src/malli/core.cljc
@@ -1705,7 +1705,7 @@
               (when (-accept walker this path options)
                 (if (or (not id) ((-boolean-fn (::walk-schema-refs options false)) id))
                   (-outer walker this path (-inner-indexed walker path children options) options)
-                  (-outer walker this path [id] options))))
+                  (-outer walker this path children #_[id] options))))
             (-properties [_] properties)
             (-options [_] options)
             (-children [_] children)
```

**Comment by opqdonut** [comment]
Discussed this with Tommi. This PR now has the more principled fix, and as a side-effect removes `mu/from-map-syntax` and `mu/to-map-syntax`.

**Comment by ikitommi** [comment]
This is really good, fixes real issue and simplifies the core 🙇, also:

Screenshot 2023-04-12 at 16 09 32

**Pull request merged**